clients/src/main/java/org/apache/kafka/clients/admin/DescribeProducersResult.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.admin;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.annotation.InterfaceStability;
+import org.apache.kafka.common.internals.KafkaFutureImpl;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+@InterfaceStability.Evolving
+public class DescribeProducersResult {
+
+    private final Map<TopicPartition, KafkaFutureImpl<PartitionProducerState>> futures;
+
+    DescribeProducersResult(Map<TopicPartition, KafkaFutureImpl<PartitionProducerState>> futures) {
+        this.futures = futures;
+    }
+
+    public KafkaFuture<PartitionProducerState> partitionResult(final TopicPartition partition) {
+        KafkaFuture<PartitionProducerState> future = futures.get(partition);
+        if (future == null) {
+            throw new IllegalArgumentException("Topic partition " + partition +
+                " was not included in the request");
+        }
+        return future;
+    }
+
+    public KafkaFuture<Map<TopicPartition, PartitionProducerState>> all() {
+        return KafkaFuture.allOf(futures.values().toArray(new KafkaFuture[0]))
+            .thenApply(nil -> {
+                Map<TopicPartition, PartitionProducerState> results = new HashMap<>(futures.size());
+                for (Map.Entry<TopicPartition, KafkaFutureImpl<PartitionProducerState>> entry : futures.entrySet()) {
+                    try {
+                        results.put(entry.getKey(), entry.getValue().get());
+                    } catch (InterruptedException | ExecutionException e) {
+                        // This should be unreachable, because allOf ensured that all the futures completed successfully.
+                        throw new KafkaException(e);
+                    }
+                }
+                return results;
+            });
+    }
+
+    public static class PartitionProducerState {
+        private final List<ProducerState> activeProducers;
+
+        public PartitionProducerState(List<ProducerState> activeProducers) {
```

Comment:
I need it public because the handler is in `admin.internals`. This might be a pattern that has to give way if we want to externalize some of the admin logic rather than packing it all into `KafkaAdminClient`. I'll try to consider some other options.

I am not sure I would call it "useless" by the way. It is helpful when you need to mock the admin client. Of course users can always mock `PartitionProducerState` as well, though it gets tedious for property classes.

Reply:
I do agree it is useful for testing but he public constructors need release cycles to deprecate. If we want to add more data in the future, the (exposed) public constructor can make trouble to us.


> I need it public because the handler is in admin.internals. This might be a pattern that has to give way if we want to externalize some of the admin logic rather than packing it all into KafkaAdminClient. I'll try to consider some other options.

I like to use annotation `@InterfaceStability.Evolving` if we all agree that it give us "rights" to break APIs in minor release. For example, add `@InterfaceStability.Evolving` to the public constructor.

Reply:
just notice that there is a class `AdminClientTestUtils` which is used to create those results for testing. Maybe we can change the modifier from public to package-private and then add a helper method to `AdminClientTestUtils` to create `PartitionProducerState`/`DescribeProducersResult` if test code needs it.